content/blog/2023-12-01-task-lifecycle.markdown

```diff
@@ -0,0 +1,470 @@
+any task creation requests. We also limit the number of tasks here, dynamically
+picking the limit based on the available memory, page size, and a fixed upper
+bound.
+
+Towards the end of the init sequence we reach `rest_init()`. In here we finally
+dynamically create our first two tasks: the init task (not to be confused with
+`init_task`), and the kthreadd task (with the double 'd'). The init task is
+(eventually) the userspace init process. We create it first to get the PID value
+1, which is relied on by a number of things in the kernel and in userspace[^5].
+The kthreadd task provides an asynchronous creation mechanism for kthreads:
+callers append their thread parameters to a dedicated list, and the kthreadd
+task spawns any entries on the list whenever it gets scheduled. Creating the
+tasks automatically put them on the scheduler run queue, and they might even
+have started automatically with preemption.
+
+[^5]: One example of the init task being special is that the kernel will not
+    allow its process to be killed. It must always have at least one thread.
+
+```c
+// init/main.c
+
+noinline void __ref __noreturn rest_init(void)
+{
+	struct task_struct *tsk;
+	int pid;
+
+	rcu_scheduler_starting();
+	/*
+	 * We need to spawn init first so that it obtains pid 1, however
+	 * the init task will end up wanting to create kthreads, which, if
+	 * we schedule it before we create kthreadd, will OOPS.
+	 */
+	pid = user_mode_thread(kernel_init, NULL, CLONE_FS);
+	/*
+	 * Pin init on the boot CPU. Task migration is not properly working
+	 * until sched_init_smp() has been run. It will set the allowed
+	 * CPUs for init to the non isolated CPUs.
+	 */
+	rcu_read_lock();
+	tsk = find_task_by_pid_ns(pid, &init_pid_ns);
+	tsk->flags |= PF_NO_SETAFFINITY;
+	set_cpus_allowed_ptr(tsk, cpumask_of(smp_processor_id()));
+	rcu_read_unlock();
+
+	numa_default_policy();
+	pid = kernel_thread(kthreadd, NULL, NULL, CLONE_FS | CLONE_FILES);
+	rcu_read_lock();
+	kthreadd_task = find_task_by_pid_ns(pid, &init_pid_ns);
+	rcu_read_unlock();
+
+	/*
+	 * Enable might_sleep() and smp_processor_id() checks.
+	 * They cannot be enabled earlier because with CONFIG_PREEMPTION=y
+	 * kernel_thread() would trigger might_sleep() splats. With
+	 * CONFIG_PREEMPT_VOLUNTARY=y the init task might have scheduled
+	 * already, but it's stuck on the kthreadd_done completion.
+	 */
+	system_state = SYSTEM_SCHEDULING;
+
+	complete(&kthreadd_done);
+
+	/*
+	 * The boot idle thread must execute schedule()
+	 * at least once to get things moving:
+	 */
+	schedule_preempt_disabled();
+	/* Call into cpu_idle with preempt disabled */
+	cpu_startup_entry(CPUHP_ONLINE);
+}
+
+```
+
+After this, it basically just enters the scheduler. We're now almost fully task
+driven, and our journey picks back up inside of the init task.
+
+> Hint: you can see where the `init_task` vs the init task is used in the
+> console. The `init_task` has PID 0, so appears as `T0` in the log line
```

Comment:
You may want to be more verbose when describing the console. Something like 'were you to watch your computer boot, you would see T0, T1' etc. Just in case it's not clear that we mean when we say 'console'.